clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -523,6 +625,17 @@ private static int parseAcks(String acksString) {
      * Implementation of asynchronously send a record to a topic.
      */
     private Future<RecordMetadata> doSend(ProducerRecord<K, V> record, Callback callback) {
+        if (transactionState != null) {
+            if (transactionState.isTransactional() && !transactionState.hasPid())
+                throw new IllegalStateException("Cannot perform a 'send' before completing a call to initTransactions when transactions are enabled.");
+
+            if (transactionState.isFenced() && transactionState.isInTransaction())
```

Comment:
Mentioned offline, but it seems we don't need the `isInTransaction` check here. No matter the state, if the producer is fenced, we should raise.